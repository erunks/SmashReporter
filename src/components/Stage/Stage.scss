.stage {
  display: flex;
  position: relative;
}

.stageCard {
  position: relative;
  margin: 0 auto;
  max-width: 325px;

  img {
    border: 1px solid black;
    height: auto;
    max-height:100%;
    max-width:100%;
    width: 150px;
  }

  .name {
    color: white;
    font-size: 12px;
    padding: 0.25em 2px;
    margin: 0;
    overflow: hidden;
    text-align: center;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;
  }

  .namePlate {
    background-color: black;
    border: 1px dashed white;
    height: 30px;
    max-width: 100%;
  }
}

.selectedStageToken {
  position: absolute;
  top: -18px;
  left: -18px;
  
  img {
    border: 2px solid black;
    border-radius: 100%;
    width: 100%;
    z-index: 99;
  }
}
