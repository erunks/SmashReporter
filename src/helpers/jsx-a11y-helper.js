export function buttonize(action) {
  return {
    role: 'button',
    onClick: action,
    onKeyDown: (event) => {
      // Handles Enter/Return pressed
      if (event.keycode === 13) {
        action(event);
      }
    }
  };
}

export default { buttonize };
