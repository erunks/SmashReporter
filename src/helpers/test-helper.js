import React from 'react';
import { render } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { IKContext } from 'imagekitio-react';
import { RecoilRoot } from 'recoil';

const wrapper = (component, options = {
  initializeState: () => {},
}) => {
  const initialStateMethod = options.initializeState;

  return (
    <IKContext
      publicKey="public_DfZPshOEvWdxzvxS8HGNWPJCo70="
      urlEndpoint="https://ik.imagekit.io/ozcerk4wii"
    >
      <RecoilRoot initializeState={ initialStateMethod }>
        { component }
      </RecoilRoot>
    </IKContext>
  );
};

const mountWrapper = (component, options) => mount(wrapper(component, options));
const shallowWrapper = (component, options) => shallow(wrapper(component, options));
const wrappedRender = (component, options = {}) => {
  const initialStateMethod = options && options.initializeState;

  const Wrapper = ({ children }) => (
    <IKContext
      publicKey="public_DfZPshOEvWdxzvxS8HGNWPJCo70="
      urlEndpoint="https://ik.imagekit.io/ozcerk4wii"
    >
      <RecoilRoot initializeState={ initialStateMethod }>
        { children }
      </RecoilRoot>
    </IKContext>
  );

  return render(component, { wrapper: Wrapper, ...options });
};


export * from '@testing-library/react';

export {
  mountWrapper,
  shallowWrapper,
  wrappedRender
};
