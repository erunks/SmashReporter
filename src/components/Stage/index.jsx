import React from 'react';
import propTypes from 'prop-types';
import { URLS } from '../../lib/imageAssets';

import './Stage.scss';

const Stage = (props) => (
    <div className="stage">
      <div className="stageCard">
        <img src={URLS.stage(props.id, props.dlc)} alt={`stage: ${props.name}`} />
        <div className="namePlate">
          <div className="name">{props.name}</div>
        </div>
      </div>
    </div>
);

const {
  bool,
  number,
  oneOfType,
  string
} = propTypes;

Stage.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  id: number.isRequired,
  name: string.isRequired
};

export default Stage;
