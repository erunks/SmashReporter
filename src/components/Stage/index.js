import React, { useRef, useState } from 'react';
import propTypes from 'prop-types';
import { useRecoilState } from 'recoil';
import { selectedStageState } from 'recoils/stage';
import { buttonize } from 'helpers/jsx-a11y-helper'
import { IKImage } from 'imagekitio-react';
import { URLS } from 'lib/imageAssets';

import './Stage.scss';

const Stage = (props) => {
  const {
    dlc,
    displayName,
    id,
    name
  } = props;

  const [selectedStage, setSelectedStage] = useRecoilState(selectedStageState);
  const [clickPostion, setClickPosition] = useState({ x: 50, y: 27 });
  const ref = useRef(null);

  const renderStageName = () => (
    <div className="namePlate">
      <div className="name">{ name }</div>
    </div>
  );

  const isStageSelected = () => id === selectedStage;
  const onClick = (event) => {
    setSelectedStage(id);
    const { left, top } = ref.current.getBoundingClientRect();
    setClickPosition({
      x: Math.abs(left - event.clientX),
      y: Math.abs(top - event.clientY)
    });
  };

  return (
    <div ref={ref} className={`stage ${isStageSelected() ? 'selected' : ''}`} { ...buttonize(onClick) }>
      <div className="stageCard">
        <IKImage
          src={URLS.stage(id, dlc)}
          alt={`stage: ${name}`}
          loading="lazy"
          lqip={{ active: true }}
        />
        { displayName && renderStageName() }
        { isStageSelected() && <div className="selectedStageToken">
          <IKImage
            path={'/assets/stage_coin.png'}
            alt="Stage select token"
            loading="lazy"
            lqip={{ active: true }}
            style={{
              position: 'relative',
              top: `${clickPostion.y}px`,
              left: `${clickPostion.x}px`,
            }}
          />
        </div> }
      </div>
    </div>
  );
};

const {
  bool,
  number,
  oneOfType,
  string
} = propTypes;

Stage.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  displayName: bool,
  id: number.isRequired,
  name: string.isRequired
};

Stage.defaultProps = {
  displayName: false
};

export default Stage;
