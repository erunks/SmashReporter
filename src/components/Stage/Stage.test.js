import React from 'react';
import { shallowWrapper } from 'helpers/test-helper';
import Stage from '.';

describe('<Stage>', () => {
  const props = {
    dlc: 0,
    id: 1,
    name: 'Random'
  };
  const container = shallowWrapper(<Stage {...props} />);

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
