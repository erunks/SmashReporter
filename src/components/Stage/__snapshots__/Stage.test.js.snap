// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stage> should match the snapshot 1`] = `"<div class=\\"stage selected\\" role=\\"button\\"><div class=\\"stageCard\\"><img alt=\\"stage: Random\\"/><div class=\\"selectedStageToken\\"><img alt=\\"Stage select token\\" style=\\"position:relative;top:27px;left:50px\\"/></div></div></div>"`;
